realm/realm-library/src/objectServer/java/io/realm/internal/network/OkHttpAuthenticationServer.java

```diff
@@ -54,7 +55,9 @@ public AuthenticateResponse loginUser(SyncCredentials credentials, URL authentic
             String requestBody = AuthenticateRequest.userLogin(credentials).toJson();
             return authenticate(authenticationUrl, requestBody);
         } catch (Exception e) {
-            return AuthenticateResponse.from(new ObjectServerError(ErrorCode.UNKNOWN, e));
+            // ConnectException is recoverable (with exponential backoff)
+            ErrorCode errorCode = (e instanceof ConnectException) ? ErrorCode.IO_EXCEPTION : ErrorCode.UNKNOWN;
+            return AuthenticateResponse.from(new ObjectServerError(errorCode, e));
```

Comment:
Wouldn't it be more relevant to have a utility method that mapped between Exception and ErrorCode? Since this is being used for both authenticate/changepassword/logout and probably future ones as well.